File: LC/Medium/347.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        freqs = [[] for _ in range(n)]
        count = Counter(nums)
        
        for key, v in count.items():
            freqs[v-1].append(key)
            
        ans = []
        for i in range(n-1,-1,-1):
            for n in freqs[i]:
                ans.append(n)
                if len(ans) == k:
                    return ans
        return ans

File: LC/C++/Medium/347.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        std::unordered_map<int, int> freq{};
        for (int num : nums) freq[num]++;

        std::vector<std::vector<int>> buckets(nums.size() + 1);
        for (auto [x, count] : freq)
            buckets[count].push_back(x);

        std::vector<int> result;
        for (int count = nums.size(); count > 0 && result.size() < k; --count) {
            for (int x : buckets[count]) {
                result.push_back(x);
                if (result.size() == k) break;
            }
        }

        return result;
    }
};